### PR TITLE
[Debugger] Added new breakpoint span resolver for C#

### PR DESCRIPTION
**Pull request by jstedfast**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/887700/

### CODE REVIEW

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/DebuggingService.cs [hunk 1]
```diff
@@ -1482,6 +1490,19 @@ public static async Task<CompletionData> GetCompletionDataAsync (StackFrame fram
 				return result;
 			return frame.GetExpressionCompletionData (exp);
 		}
+
+		public static Task<Span> GetBreakpointSpanAsync (ITextDocument document, int position, CancellationToken cancellationToken = default (CancellationToken))
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/DebuggingService.cs]: Where is this used?

Comment (jstedfast) [reply main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/DebuggingService.cs]: Right here: https://github.com/mono/monodevelop/pull/7787/commits/a339be94d3cfbec72f33c8ac9ffdae846638c286#diff-eee1a8eea31b7ee727ff9a017a563dc6R78

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/IBreakpointSpanResolver.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+using System.Threading.Tasks;
+
+using Microsoft.VisualStudio.Text;
+
+namespace MonoDevelop.Debugger
+{
+	public interface IBreakpointSpanResolver
+	{
+		Task<Span> GetBreakpointSpanAsync (ITextBuffer buffer, int position, CancellationToken cancellationToken);
+	}
```
Comment (slluis) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/IBreakpointSpanResolver.cs]: Please add some documentation to the interface and the method.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			var document = buffer.AsTextContainer ().GetOpenDocumentInCurrentContext ();
+			var tree = await document.GetSyntaxTreeAsync (cancellationToken);
+
+			BreakpointSpans.TryGetBreakpointSpan (tree, position, cancellationToken, out var span);
+
+			return new Span (span.Start, span.Length);
+		}
+	}
+
+	[ExportDocumentControllerExtension (MimeType = "*")]
```
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs]: If it is specific to c#, then specify MimeType="text/x-csharp"

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	class DebuggerBreakpointSpanController : DocumentControllerExtension
+	{
+		public override Task<bool> SupportsController (DocumentController controller)
+		{
+			return Task.FromResult (controller.GetContent<ITextBuffer> () != null);
+		}
+
+		protected override object OnGetContent (Type type)
+		{
+			if (typeof (IBreakpointSpanResolver).IsAssignableFrom (type))
```
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs]: You can also implement `IBreakpointSpanResolver` directly in `DebuggerBreakpointSpanController`, in which case it is not necessary to implement OnGetContent

Comment (jstedfast) [reply main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs]: done

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs (hunk 3 above)
Comment (jstedfast) [reply main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs]: done

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/IBreakpointSpanResolver.cs (hunk 2 above)
Comment (jstedfast) [reply main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/IBreakpointSpanResolver.cs]: done

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+using Microsoft.VisualStudio.Text;
+using Microsoft.CodeAnalysis.CSharp.EditAndContinue;
+
+using MonoDevelop.Debugger;
+using MonoDevelop.Ide.Gui.Documents;
+
+namespace MonoDevelop.CSharp.Debugger
+{
+	[ExportDocumentControllerExtension (MimeType = "text/x-csharp")]
+	public class CSharpBreakpointSpanResolver : DocumentControllerExtension, IBreakpointSpanResolver
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Debugger/CSharpBreakpointSpanResolver.cs]: Does this need to be public?

**Pull request merged**